### PR TITLE
Made ScenarioWorld and related stuff thread safe

### PR DESCRIPTION
**Pull request by gfusee**
I replace non Send and Sync types Rc and RefCell by their equivalent Arc and Mutex
I also added Send + Sync constraints where needed

I changed some names accordingly to the changes made

### CODE REVIEW

**Comment by gfusee** [comment]
This is a draft bc I'm gonna write at least one test that ensures that a scenario can be run in a multithreading context (probably using tokio::spawn)

The code itself will not change

**Review by andrei-marinica — DISMISSED** [review]

File: framework/scenario/src/facade/scenario_world.rs [hunk 1]
```diff
@@ -145,7 +145,7 @@ impl ScenarioWorld {
         self.get_mut_debugger_backend()
             .vm_runner
             .contract_map_ref
-            .borrow_mut()
+            .borrow()
```
Comment (andrei-marinica) [inline framework/scenario/src/facade/scenario_world.rs:148]: Why is `borrow_mut` no longer necessary? This should be a mutable operation.

File: framework/scenario/src/whitebox_legacy/contract_obj_wrapper.rs [hunk 2]
```diff
@@ -270,7 +271,7 @@ impl BlockchainStateWrapper {
                 .get_mut_debugger_backend()
                 .vm_runner
                 .contract_map_ref
-                .borrow_mut()
+                .borrow()
```
Comment (andrei-marinica) [inline framework/scenario/src/whitebox_legacy/contract_obj_wrapper.rs:274]: Why is borrow_mut no longer necessary? This should be a mutable operation.

File: framework/scenario/tests/contract_without_macros.rs [hunk 3]
```diff
@@ -286,7 +286,7 @@ mod sample_adder {
 
     impl<A> multiversx_sc::contract_base::CallableContract for ContractObj<A>
     where
-        A: multiversx_sc::api::VMApi,
+        A: multiversx_sc::api::VMApi + Send + Sync,
```
Comment (andrei-marinica) [inline framework/scenario/tests/contract_without_macros.rs:289]: Ideally, the contract shouldn't know about Send/Sync. Doesn't it work without this?

File: framework/scenario/tests/contract_without_macros.rs [hunk 4]
```diff
@@ -299,7 +299,7 @@ mod sample_adder {
     pub struct ContractBuilder;
 
     impl multiversx_sc::contract_base::CallableContractBuilder for ContractBuilder {
-        fn new_contract_obj<A: multiversx_sc::api::VMApi>(
+        fn new_contract_obj<A: multiversx_sc::api::VMApi + Send + Sync>(
```
Comment (andrei-marinica) [inline framework/scenario/tests/contract_without_macros.rs:302]: Ideally, the contract shouldn't know about Send/Sync. Doesn't it work without this?

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

**Review by alyn509 — APPROVED** [review]

File: framework/scenario/src/facade/scenario_world.rs (hunk 1 above)
Comment (gfusee) [reply framework/scenario/src/facade/scenario_world.rs:148]: with Mutex there is only a lock mechanism, no difference between borrow and borrow_mut like there is on RefCell, I'll rename the function to reflect that

File: framework/scenario/src/whitebox_legacy/contract_obj_wrapper.rs (hunk 2 above)
Comment (gfusee) [reply framework/scenario/src/whitebox_legacy/contract_obj_wrapper.rs:274]: with Mutex there is only a lock mechanism, no difference between borrow and borrow_mut like there is on RefCell, I'll rename the function to reflect that

File: framework/scenario/tests/contract_without_macros.rs (hunk 3 above)
Comment (gfusee) [reply framework/scenario/tests/contract_without_macros.rs]: Done by adding the constraint Send + Sync on traits VMApi and VMHooksApiBackend

File: framework/scenario/src/facade/scenario_world.rs (hunk 1 above)
Comment (gfusee) [reply framework/scenario/src/facade/scenario_world.rs]: done by renaming the function "borrow" to "lock" to better match the RefCell -> Mutex conversion

"lock" = both old "borrow" and "borrow_mut" functions

File: framework/scenario/src/whitebox_legacy/contract_obj_wrapper.rs (hunk 2 above)
Comment (gfusee) [reply framework/scenario/src/whitebox_legacy/contract_obj_wrapper.rs]: done by renaming the function "borrow" to "lock" to better match the RefCell -> Mutex conversion

"lock" = both old "borrow" and "borrow_mut" functions

**Review by andrei-marinica — APPROVED** [review]

**Pull request merged**